File: packages/faro-android-sdk/src/main/kotlin/com/edkimmel/faro/transport/UploadConditions.kt
package com.edkimmel.faro.transport

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.PowerManager

/**
 * Checks system conditions before attempting network uploads.
 * Modeled after Datadog's approach: battery level, charging state, power-save mode.
 */
internal object UploadConditions {

    private const val LOW_BATTERY_THRESHOLD = 0.10f

    enum class Blocker(val reason: String) {
        LOW_BATTERY("Battery level below threshold"),
        POWER_SAVE_MODE("Power save mode enabled")
    }

    /**
     * Returns a list of conditions currently blocking uploads.
     * Empty list means uploads are allowed.
     */
    fun currentBlockers(context: Context): List<Blocker> {
        val blockers = mutableListOf<Blocker>()

        checkBattery(context, blockers)
        checkPowerSaveMode(context, blockers)

        return blockers
    }

    private fun checkBattery(context: Context, blockers: MutableList<Blocker>) {
        try {
            val batteryStatus: Intent? = IntentFilter(Intent.ACTION_BATTERY_CHANGED).let { filter ->
                context.registerReceiver(null, filter)
            }
            if (batteryStatus == null) return

            val status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
            val isCharging = status == BatteryManager.BATTERY_STATUS_CHARGING
                    || status == BatteryManager.BATTERY_STATUS_FULL

            if (isCharging) return

            val level = batteryStatus.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
            val scale = batteryStatus.getIntExtra(BatteryManager.EXTRA_SCALE, -1)

            if (level >= 0 && scale > 0) {
                val batteryPct = level.toFloat() / scale.toFloat()
                if (batteryPct < LOW_BATTERY_THRESHOLD) {
                    blockers.add(Blocker.LOW_BATTERY)
                }
            }
        } catch (_: Exception) {
            // If we can't read battery state, allow uploads
        }
    }

    private fun checkPowerSaveMode(context: Context, blockers: MutableList<Blocker>) {
        try {
            val powerManager = context.getSystemService(Context.POWER_SERVICE) as? PowerManager
            if (powerManager?.isPowerSaveMode == true) {
                blockers.add(Blocker.POWER_SAVE_MODE)
            }
        } catch (_: Exception) {
            // If we can't read power state, allow uploads
        }
    }
}

File: packages/faro-ios-sdk/Sources/FaroSDK/Transport/UploadConditions.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Checks system conditions before attempting network uploads.
/// Modeled after Datadog's approach: battery, low-power mode, network reachability.
internal struct UploadConditions {

    enum Blocker: String {
        case lowBattery = "Battery level below threshold"
        case lowPowerMode = "Low power mode enabled"
    }

    /// Minimum battery level (fraction 0–1) above which uploads are allowed.
    static let minimumBatteryLevel: Float = 0.1

    /// Returns blockers preventing upload. Empty array means upload is allowed.
    static func currentBlockers() -> [Blocker] {
        var blockers: [Blocker] = []

        #if canImport(UIKit) && !os(watchOS)
        checkBattery(&blockers)
        checkLowPowerMode(&blockers)
        #endif

        return blockers
    }

    #if canImport(UIKit) && !os(watchOS)
    private static func checkBattery(_ blockers: inout [Blocker]) {
        let device = UIDevice.current
        let wasMonitoring = device.isBatteryMonitoringEnabled
        device.isBatteryMonitoringEnabled = true
        defer {
            if !wasMonitoring { device.isBatteryMonitoringEnabled = false }
        }

        let state = device.batteryState
        let level = device.batteryLevel

        // If charging, full, or unknown (e.g. simulator), allow uploads
        if state == .charging || state == .full || state == .unknown {
            return
        }

        // On battery power — check level
        if level >= 0 && level < minimumBatteryLevel {
            blockers.append(.lowBattery)
        }
    }

    private static func checkLowPowerMode(_ blockers: inout [Blocker]) {
        if ProcessInfo.processInfo.isLowPowerModeEnabled {
            blockers.append(.lowPowerMode)
        }
    }
    #endif
}
